### PR TITLE
Implement DROP SUBSCRIPTION in 3 requests

### PR DESCRIPTION
**Pull request by BaurzhanSakhariev**
Follow up to https://github.com/crate/crate/pull/11897#issuecomment-968953154

Replacement of https://github.com/crate/crate/pull/11937

- [x] stop tracking of remote shards (by CLOSE TABLE)
- [x] remove retention leases on the remote cluster for each table/index (by CLOSE TABLE)
- [x] remove the REPLICATION_SUBSCRIPTION_NAME index setting to turn the tables into normal tables
- [x] update doc info for all subscription tables to make them normal after DROP SUBSCRIPTION (by OPEN TABLE)

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: server/src/test/java/io/crate/integrationtests/LogicalReplicationITest.java [hunk 1]
```diff
@@ -437,4 +437,37 @@ public void test_subscription_state_order() throws Exception {
...
+            OperationOnInaccessibleRelationException.class,
+            "The relation \"doc.t2\" doesn't support or allow INSERT operations."
+        );
+
+        executeOnSubscriber("DROP SUBSCRIPTION sub1 ");
+
+        var response = executeOnSubscriber("INSERT INTO doc.t1 (id) VALUES(3)");
+        assertThat(response.rowCount(), is(1L));
+
+        Thread.sleep(1000);
```
Comment (BaurzhanSakhariev) [inline server/src/test/java/io/crate/integrationtests/LogicalReplicationITest.java:469]: Passes with sleep and without it fails with message:
`Expected valid sequence number for replicated op but was unassigned`
I'm wondering if it's not a bug and I need to use better way instead of sleep.

File: server/src/test/java/io/crate/integrationtests/LogicalReplicationITest.java [hunk 2]
```diff
@@ -437,4 +437,37 @@ public void test_subscription_state_order() throws Exception {
...
+            OperationOnInaccessibleRelationException.class,
+            "The relation \"doc.t2\" doesn't support or allow INSERT operations."
+        );
+
+        executeOnSubscriber("DROP SUBSCRIPTION sub1 ");
+
+        var response = executeOnSubscriber("INSERT INTO doc.t1 (id) VALUES(3)");
+        assertThat(response.rowCount(), is(1L));
+
+        Thread.sleep(500);
```
Comment (BaurzhanSakhariev) [inline server/src/test/java/io/crate/integrationtests/LogicalReplicationITest.java:469]: Passes with sleep and without it fails with message:
Expected valid sequence number for replicated op but was unassigned
I'm wondering if it's not a bug and I need to use better way instead of sleep.

Comment (seut) [reply server/src/test/java/io/crate/integrationtests/LogicalReplicationITest.java:469]: This indicates a bug inside the chain as the failure tells us that the wrong engine is still used for the table. Thus the close->remove-setting->open did not finish yet.

File: server/src/main/java/io/crate/metadata/table/Operation.java [hunk 3]
```diff
@@ -64,7 +64,7 @@
     public static final EnumSet<Operation> METADATA_DISABLED_OPERATIONS = EnumSet.of(READ, UPDATE, INSERT, DELETE,
         ALTER_BLOCKS, ALTER_OPEN_CLOSE, ALTER_REROUTE, REFRESH, SHOW_CREATE, OPTIMIZE);
     public static final EnumSet<Operation> LOGICAL_REPLICATED = EnumSet.of(
-        READ, ALTER_BLOCKS, ALTER_REROUTE, OPTIMIZE, REFRESH, COPY_TO, SHOW_CREATE);
+        READ, ALTER_BLOCKS, ALTER_REROUTE, OPTIMIZE, REFRESH, COPY_TO, SHOW_CREATE, ALTER_OPEN_CLOSE);
```
Comment (seut) [inline server/src/main/java/io/crate/metadata/table/Operation.java:67]: This would allow a user to close/open a table with is currently replicating. Is this really needed? We shouldn't allow that (at least not now) as it would technical equal to pausing a subscription as trackers are stopped. If the gap get larger and thus retention leases time out, the shard trackers cannot replicate anymore and we'd have to a) bail out or b) restore the table from scratch using the repository logic.

Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/metadata/table/Operation.java:67]: I think initially I added it only because I did ALTER TABLE OPEN/CLOSE manually in tests - but now I recall discussion about hints on failure - quote from slack

```
first step (close tables) fails, the user can safely re-run DROP SUBSCRIPTION.
2nd step (drop setting + subscription) failes, same as ^^
3rd step (open tables) fails, user should be able to manually open the related tables using ALTER TABLE OPEN 
```

for the 3-step failure we need at least let OPEN - for now we don't have such granularity between open/close - I can think of some tricky condition inside `maybeRaiseBlockedException`if a table has LOGICAL_REPLICATED permissions, we need to treat close/open as `only open`

Comment (seut) [reply server/src/main/java/io/crate/metadata/table/Operation.java:67]: Ah right, yes for that we need to allow `open`. Maybe we should replace the operation with dedicated `ALTER_OPEN` + `ALTER_CLOSE` ones?

Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/metadata/table/Operation.java]: done in https://github.com/crate/crate/pull/12112/commits/faead852afeb7e06b3dafa08994a9fd7438e39f5

**Comment by BaurzhanSakhariev** [comment]
Hi @seut, I addressed comments (my own about hints and splitting open.close). Also committed a fix (last commit) - not sure though, is it future proof - just aligned priority with [ES](https://github.com/elastic/elasticsearch/commit/8e5dd20efb667d272fac4b2a151f8c09f0284024#diff-18093c3700f3a8d43f579448b442924c7606856877de750e41c7a9e078e955eeR127)

**Comment by BaurzhanSakhariev** [comment]
./gradlew :server:test -Dtests.seed=70135F98DF76A15 -Dtests.class=io.crate.integrationtests.PostgresITest -Dtests.method="testExecuteBatchWithDifferentStatements" -Dtests.locale=dz -Dtests.timezone=Asia/Rangoon


not related

File: docs/general/information-schema.rst [hunk 4]
```diff
@@ -124,7 +124,7 @@ number of replicas.
     | sys                | summits                 | BASE TABLE |             NULL | NULL               |
     | sys                | users                   | BASE TABLE |             NULL | NULL               |
     +--------------------+-------------------------+------------+------------------+--------------------+
-    SELECT 56 rows in set (... sec)
+    SELECT 57 rows in set (... sec)
```
Comment (seut) [inline docs/general/information-schema.rst:127]: curious: how is this related to this PR?

File: server/src/main/java/io/crate/execution/ddl/tables/CloseTableRequest.java [hunk 5]
```diff
@@ -23,40 +23,60 @@
...
     }
 
     @Override
     public void writeTo(StreamOutput out) throws IOException {
         super.writeTo(out);
-        table.writeTo(out);
+        if (out.getVersion().before(Version.fromString("4.8.0"))) {
+            // Before 4.8.0 request was used only for closing table with non-null RelationName field.
+            // In this branch it's guaranteed that we are not passing empty list
+            tables.get(0).writeTo(out);
```
Comment (seut) [inline server/src/main/java/io/crate/execution/ddl/tables/CloseTableRequest.java:68]: please add an assertion to ensure there is at least 1 table. NPE or such in the streaming are hard to grasp otherwise.

File: server/src/main/java/io/crate/execution/ddl/tables/OpenCloseTableOrPartitionRequest.java [hunk 6]
```diff
@@ -58,16 +57,37 @@ boolean isOpenTable() {
...
     }
 
     @Override
     public void writeTo(StreamOutput out) throws IOException {
         super.writeTo(out);
-        relationName.writeTo(out);
+        if (out.getVersion().before(Version.fromString("4.8.0"))) {
+            // Before 4.8.0 request was used only for closing table with non-null RelationName field.
+            // In this branch it's guaranteed that we are not passing empty list
+            tables.get(0).writeTo(out);
```
Comment (seut) [inline server/src/main/java/io/crate/execution/ddl/tables/OpenCloseTableOrPartitionRequest.java:79]: same here, please protect by an assertions.

File: server/src/main/java/io/crate/execution/ddl/tables/TransportCloseTable.java [hunk 7]
```diff
@@ -142,35 +146,44 @@ protected void masterOperation(CloseTableRequest request,
         clusterService.submitStateUpdateTask("add-block-close-table", new AddCloseBlocksTask(listener, request));
     }
 
+    public static String[] getIndices(List<RelationName> tables, @Nullable String partition) {
+        return partition != null ? new String[] {partition} : tables.stream()
+            .map(t -> t.indexNameOrAlias())
+            .toArray(String[]::new);
```
Comment (seut) [inline server/src/main/java/io/crate/execution/ddl/tables/TransportCloseTable.java:152]: I'd suggest to return `Indices[]` and use `indexNameExpressionResolver.concreteIndices()` here inside to avoid creating a string array for the partition.

File: server/src/main/java/io/crate/metadata/cluster/AbstractOpenCloseTableClusterStateTaskExecutor.java [hunk 8]
```diff
@@ -83,19 +86,23 @@ public PartitionName partitionName() {
     }
 
     protected Context prepare(ClusterState currentState, OpenCloseTableOrPartitionRequest request) {
-        RelationName relationName = request.tableIdent();
         String partitionIndexName = request.partitionIndexName();
         Metadata metadata = currentState.metadata();
-        String indexToResolve = partitionIndexName != null ? partitionIndexName : relationName.indexNameOrAlias();
+        String[] indexToResolve = partitionIndexName != null ? new String[] {partitionIndexName} :
+            request.tables().stream().map(t -> t.indexNameOrAlias()).toArray(String[]::new);
```
Comment (seut) [inline server/src/main/java/io/crate/metadata/cluster/AbstractOpenCloseTableClusterStateTaskExecutor.java:92]: Same here, can we avoid the string array for a single partition?

File: server/src/main/java/io/crate/replication/logical/plan/DropSubscriptionPlan.java [hunk 9]
```diff
@@ -51,8 +65,61 @@ public void executeOrFail(DependencyCarrier dependencies,
...
+
+
+        // Step 3
+        // Reopen table to update table engine to normal (Operation will be reset back to ALL after step 2)
+        actions.add(new ChainableAction<>(
+            () -> dependencies.alterTableOperation()
+                .executeAlterTableOpenClose(tables, true, null)
+                .exceptionally(error -> {
+                    throw new IllegalStateException(
+                        "Couldn't reopen tables, please run command ALTER TABLE OPEN for all subscribed tables."
```
Comment (seut) [inline server/src/main/java/io/crate/replication/logical/plan/DropSubscriptionPlan.java:115]: Ca we maybe include the  tables inside the error message?

**Review by seut — COMMENTED** [review]
Looks good in general, added some comments about details.

File: docs/general/information-schema.rst (hunk 4 above)
Comment (BaurzhanSakhariev) [reply docs/general/information-schema.rst:127]: not related, but I decided to fix it anyway - we have this failure in the main LR [branch](https://github.com/crate/crate/pull/11832/commits) -

File: server/src/main/java/io/crate/replication/logical/plan/DropSubscriptionPlan.java (hunk 9 above)
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/replication/logical/plan/DropSubscriptionPlan.java:115]: Should I add some check for tables size? So that it't not some unreadable long text if it's FOR ALL TABLES and number of tables is big.

File: server/src/main/java/io/crate/execution/ddl/tables/TransportCloseTable.java [hunk 10]
```diff
@@ -142,35 +146,44 @@ protected void masterOperation(CloseTableRequest request,
...
-            updatedState = currentState;
-        } else {
-            Metadata.Builder metadata = Metadata.builder(currentState.metadata());
-            metadata.put(closePartitionTemplate(templateMetadata));
-            updatedState = ClusterState.builder(currentState).metadata(metadata).build();
-        }
+            IndexTemplateMetadata templateMetadata = target.templateMetadata();
+            if (templateMetadata == null) {
+                // Update only on the first iteration otherwise just use value from the previous iteration
+                updatedState = updatedState == null ? currentState : updatedState;
```
[the remark this answers is not in the document]
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/execution/ddl/tables/TransportCloseTable.java]: yep, makes it simpler, thanks

File: server/src/main/java/io/crate/execution/ddl/tables/TransportCloseTable.java (hunk 7 above)
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/execution/ddl/tables/TransportCloseTable.java:152]: this method used in 2 other blocks expecting varargs.
```
@Override
protected ClusterBlockException checkBlock(CloseTableRequest request, ClusterState state) {
    return state.blocks().indicesBlockedException(ClusterBlockLevel.METADATA_WRITE,
        indexNameExpressionResolver.concreteIndexNames(state, STRICT_INDICES_OPTIONS, getIndices(request.tables(), request.partition())));
}
```
if I make it return `Index[]` I would need to almost copy that code in those 2 places.

Comment (seut) [reply server/src/main/java/io/crate/execution/ddl/tables/TransportCloseTable.java:152]: ok right, varargs will convert it anyhow

File: server/src/main/java/io/crate/replication/logical/plan/DropSubscriptionPlan.java (hunk 9 above)
Comment (seut) [reply server/src/main/java/io/crate/replication/logical/plan/DropSubscriptionPlan.java]: good point 👍

**Review by seut — APPROVED** [review]
👍 Please keep the unrelated documentation related changes in a dedicated commit.

**Comment by BaurzhanSakhariev** [comment]
`crate.client.exceptions.ConnectionError: No more Servers available`

seems to be infrastructural issue. retest this please
